File: services/c-hex-uds/src/domain/usecase.c
// c-hex-http/src/domain/usecase.c: the use case implementation.
#include "usecase.h"
#include <string.h>

void usecase_verify(const TokenVerifierPort *port, const char *token, VerificationResult *out) {
    out->valid = 0;
    out->subject[0] = 0;
    out->code = DCODE_INVALID_TOKEN;

    if (!token || !*token) {
        out->code = DCODE_BAD_REQUEST;
        return;
    }
    if (!port || !port->verify) {
        out->code = DCODE_INVALID_TOKEN;
        return;
    }
    int err = 0;
    port->verify(token, out->subject, sizeof(out->subject), &err);
    if (err == 0) {
        out->valid = 1;
        out->code = DCODE_OK;
    } else if (err == -2) {
        out->code = DCODE_BAD_REQUEST;
    } else {
        out->code = DCODE_INVALID_TOKEN;
    }
}

File: services/rust-hex-uds/src/domain/usecase.rs
use crate::domain::port::TokenVerifierPort;
use crate::domain::result::{DomainCode, VerificationResult};

pub fn verify(port: &dyn TokenVerifierPort, token: &str) -> VerificationResult {
    if token.is_empty() {
        return VerificationResult { valid: false, subject: String::new(), code: DomainCode::BadRequest };
    }
    match port.verify(token) {
        Ok(sub) => VerificationResult { valid: true, subject: sub, code: DomainCode::Ok },
        Err(e) => VerificationResult { valid: false, subject: String::new(), code: e.to_domain() },
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::domain::port::{RepoError, RepoResult};

    struct Fake(Result<String, RepoError>);
    impl TokenVerifierPort for Fake { fn verify(&self, _: &str) -> RepoResult { self.0.clone() } }

    #[test] fn empty() { let p = Fake(Ok("x".into())); assert_eq!(verify(&p, "").code, DomainCode::BadRequest); }
    #[test] fn ok() { let p = Fake(Ok("a".into())); let r = verify(&p, "t"); assert!(r.valid); }
    #[test] fn invalid() { let p = Fake(Err(RepoError::Invalid)); assert_eq!(verify(&p, "t").code, DomainCode::InvalidToken); }
}
